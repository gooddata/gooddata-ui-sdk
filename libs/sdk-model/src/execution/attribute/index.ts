// (C) 2019-2023 GoodData Corporation
import { Identifier, isUriRef, ObjRef, isIdentifierRef } from "../../objRef/index.js";
import isEmpty from "lodash/isEmpty.js";
import { invariant } from "ts-invariant";

/**
 * Attribute is our nomenclature for 'dimension' as typically used in multi-dimensional BI modeling.
 *
 * @remarks
 * The attribute specifies on what dimension to slice/dice the measures.
 *
 * In GoodData, each attribute can have multiple different display forms. Display forms essentially provide different
 * representations for elements in the dimension.
 *
 * @public
 */
export interface IAttributeBody {
    /**
     * User-assigned id of the attribute. The value is used to cross-reference model elements that are part
     * of an execution.
     *
     * The constraint is that the local identifiers of all measures and attributes in an execution definition MUST
     * be unique.
     */
    localIdentifier: Identifier;

    /**
     * Reference to display form to use for element representations.
     *
     * The display form can be referenced by either URL of the display form resource on backend OR by
     * unique, backend-recognized, identifier of the display form.
     *
     * Note: specifying display forms by URI is discouraged and WILL be deprecated in the future. The
     * specification of URL has several drawbacks:
     *
     * -  Display form URLs are tied to particular analytical workspace; this makes any application that uses
     *    URL-specified display forms workspace-specific.
     * -  Display form URLs are not supported by all GoodData backends == this makes any application that uses
     *    URL-specified display form backend-specific
     */
    displayForm: ObjRef;

    /**
     * This is user-assigned alias of the attribute; it is opaque for the backend. When backend returns
     * metadata about execution results, it WILL include this user-assigned alias in the metadata.
     */
    alias?: string;

    /**
     * Indicates whether to show all values of given attribute even if the data bound to those values
     * are not available
     */
    showAllValues?: boolean;
}

/**
 * Attribute is our nomenclature for 'dimension' as typically used in multi-dimensional BI modeling.
 *
 * @remarks
 * The attribute specifies on what dimension to slice/dice the measures.
 *
 * In GoodData, each attribute can have multiple different display forms. Display forms essentially provide different
 * representations for elements in the dimension.
 *
 * @public
 */
export interface IAttribute {
    attribute: IAttributeBody;
}

/**
 * Type guard checking whether object is an instance of IAttribute.
 *
 * @public
 */
export function isAttribute(obj: unknown): obj is IAttribute {
    return !isEmpty(obj) && (obj as IAttribute).attribute !== undefined;
}

/**
 * Defines function signature for measure predicates.
 *
 * @public
 */
export type AttributePredicate = (attribute: IAttribute) => boolean;

/**
 * Predicate that returns true for any attribute.
 *
 * @public
 */
export const anyAttribute: AttributePredicate = (_) => true;

/**
 * Factory function for attribute predicate which evaluates true for attributes that match particular ID.
 *
 * @public
 */
export const idMatchAttribute: (id: string) => AttributePredicate = (id) => (attr) =>
    attr.attribute.localIdentifier === id;

//
//
//

/**
 * Gets local identifier of an attribute.
 *
 * @remarks
 * For convenience and fluency, this function accepts both attribute object and identifier.
 *
 * @param attributeOrId - attribute to work with or the identifier
 * @returns value of local identifier
 * @public
 */
export function attributeLocalId(attributeOrId: IAttribute | Identifier): string {
    invariant(attributeOrId, "attribute must be specified");

    return typeof attributeOrId === "string" ? attributeOrId : attributeOrId.attribute.localIdentifier;
}

/**
 * Gets URI of attribute's display form to use and get attribute element values from.
 *
 * @param attribute - attribute to work with, may be undefined == result is undefined
 * @returns display form URI as string, undefined if display form not specified using URI
 * @public
 */
export function attributeUri(attribute: IAttribute): string | undefined {
    invariant(attribute, "attribute must be specified");

    return isUriRef(attribute.attribute.displayForm) ? attribute.attribute.displayForm.uri : undefined;
}

/**
 * Gets identifier of attribute's display form to use and get attribute element values from.
 *
 * @param attribute - attribute to work with, may be undefined == result is undefined
 * @returns display form identifier as string, undefined if display for not specified using identifier
 * @public
 */
export function attributeIdentifier(attribute: IAttribute): string | undefined {
    invariant(attribute, "attribute must be specified");

    return isIdentifierRef(attribute.attribute.displayForm)
        ? attribute.attribute.displayForm.identifier
        : undefined;
}

/**
 * Gets an attribute alias.
 *
 * @param attribute - attribute to work with
 * @returns value of attribute alias
 * @public
 */
export function attributeAlias(attribute: IAttribute): string | undefined {
    invariant(attribute, "attribute must be specified");

    return attribute.attribute.alias;
}

/**
 * Gets an attribute show all values property.
 *
 * @param attribute - attribute to work with
 * @returns value of attribute show all values property
 * @public
 */
export function attributeShowAllValues(attribute: IAttribute): boolean | undefined {
    invariant(attribute, "attribute must be specified");

    return attribute.attribute.showAllValues;
}

/**
 * Gets an attribute display form object ref.
 *
 * @param attribute - attribute to work with
 * @returns value of attribute display form object ref
 * @public
 */
export function attributeDisplayFormRef(attribute: IAttribute): ObjRef {
    invariant(attribute, "attribute must be specified");

    return attribute.attribute.displayForm;
}

/**
 * Given list of attributes, returns first-found attribute matching the provided predicate.
 *
 * @remarks
 * If no predicate is provided, then the function defaults to anyAttribute predicate - meaning first found attribute
 * will be returned.
 *
 * This function also provides convenience to find attribute by its local identifier - if you pass predicate as
 * string the function will automatically create idMatchAttribute predicate.
 *
 * @param attributes - list of attributes to work with, must be specified
 * @param idOrFun - attribute identifier or instance of AttributePredicate; {@link anyAttribute} predicate is default
 * @public
 */
export function attributesFind(
    attributes: IAttribute[],
    idOrFun: string | AttributePredicate = anyAttribute,
): IAttribute | undefined {
    invariant(attributes, "attributes must be specified");

    if (!attributes.length) {
        return;
    }

    const predicate = typeof idOrFun === "string" ? idMatchAttribute(idOrFun) : idOrFun;

    return attributes.find(predicate);
}
