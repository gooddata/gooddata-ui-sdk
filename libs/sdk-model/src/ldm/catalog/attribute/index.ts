// (C) 2019-2020 GoodData Corporation
import isEmpty from "lodash/isEmpty.js";
import { IAttributeMetadataObject } from "../../metadata/attribute/index.js";
import { IAttributeDisplayFormMetadataObject } from "../../metadata/attributeDisplayForm/index.js";
import { IGroupableCatalogItemBase } from "../group/index.js";

/**
 * Type representing catalog attribute
 *
 * @public
 */
export interface ICatalogAttribute extends IGroupableCatalogItemBase {
    /**
     * Catalog item type
     */
    type: "attribute";

    /**
     * Attribute metadata object that catalog attribute represents
     */
    attribute: IAttributeMetadataObject;

    /**
     * Default display form of the attribute
     */
    defaultDisplayForm: IAttributeDisplayFormMetadataObject;

    /**
     * Display forms of the attribute
     */
    displayForms: IAttributeDisplayFormMetadataObject[];

    /**
     * Attribute's display forms that contain geo pins (lat; lng) pairs.
     */
    geoPinDisplayForms: IAttributeDisplayFormMetadataObject[];
}

/**
 * Type guard checking whether the provided object is a {@link ICatalogAttribute}
 *
 * @public
 */
export function isCatalogAttribute(obj: unknown): obj is ICatalogAttribute {
    return !isEmpty(obj) && (obj as ICatalogAttribute).type === "attribute";
}
