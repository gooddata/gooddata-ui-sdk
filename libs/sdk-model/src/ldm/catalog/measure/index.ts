// (C) 2019-2020 GoodData Corporation
import isEmpty from "lodash/isEmpty.js";
import { IGroupableCatalogItemBase } from "../group/index.js";
import { IMeasureMetadataObject } from "../../metadata/measure/index.js";

/**
 * Type representing catalog measure
 *
 * @public
 */
export interface ICatalogMeasure extends IGroupableCatalogItemBase {
    /**
     * Catalog item type
     */
    type: "measure";

    /**
     * Measure metadata object that catalog measure represents
     */
    measure: IMeasureMetadataObject;
}

/**
 * Type guard checking whether the provided object is a {@link ICatalogMeasure}
 *
 * @public
 */
export function isCatalogMeasure(obj: unknown): obj is ICatalogMeasure {
    return !isEmpty(obj) && (obj as ICatalogMeasure).type === "measure";
}
