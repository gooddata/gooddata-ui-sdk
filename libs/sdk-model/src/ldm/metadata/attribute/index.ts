// (C) 2019-2022 GoodData Corporation
import { IMetadataObject, isMetadataObject } from "../types.js";
import { IAttributeDisplayFormMetadataObject } from "../attributeDisplayForm/index.js";
import { ObjRef } from "../../../objRef/index.js";

/**
 * Attribute metadata object
 *
 * @public
 */
export interface IAttributeMetadataObject extends IMetadataObject {
    type: "attribute";
    /**
     * A reference to the attribute displayForm that represents implicit drill down step
     *
     * @remarks
     * Drilling of this type will be available in any report/dashboard where this attribute will be present.
     * This will be performed on attribute headers and attribute element headers. These will be defined in LDM.
     */
    drillDownStep?: ObjRef;

    /**
     * A reference to the attribute displayForm that represents implicit drill to url from attribute value
     *
     * @remarks
     * Drilling of this type will be available in any report/dashboard where this attribute will be present.
     * This will be performed on attribute headers and attribute element headers. These will be defined in LDM.
     */
    drillToAttributeLink?: ObjRef;

    /**
     * Display forms of the attribute
     */
    displayForms: IAttributeDisplayFormMetadataObject[];
}

/**
 * Tests whether the provided object is of type {@link IAttributeMetadataObject}.
 *
 * @param obj - object to test
 * @public
 */
export function isAttributeMetadataObject(obj: unknown): obj is IAttributeMetadataObject {
    return isMetadataObject(obj) && obj.type === "attribute";
}
