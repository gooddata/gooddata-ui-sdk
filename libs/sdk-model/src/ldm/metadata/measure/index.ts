// (C) 2019-2024 GoodData Corporation
import { IMetadataObject, IMetadataObjectDefinition, isMetadataObject } from "../types.js";
import { IAuditable } from "../../../base/metadata.js";

/**
 * @public
 */
export interface IMeasureMetadataObjectBase {
    type: "measure";

    /**
     * Measure MAQL expression
     */
    expression: string;

    /**
     * Measure formatting
     * Prefer set format value, if the format is empty string backend implementation-dependent default will be used.
     */
    format: string;

    /**
     * Measure is locked for editing
     */
    isLocked?: boolean;

    /**
     * Measure tags.
     */
    tags?: string[];
}

/**
 * Measure metadata object
 *
 * @public
 */
export type IMeasureMetadataObject = IMetadataObject & IMeasureMetadataObjectBase & IAuditable;

/**
 * Measure metadata object definition
 *
 * @public
 */
export type IMeasureMetadataObjectDefinition = IMetadataObjectDefinition & IMeasureMetadataObjectBase;

/**
 * Tests whether the provided object is of type {@link IMeasureMetadataObject}.
 *
 * @param obj - object to test
 * @public
 */
export function isMeasureMetadataObject(obj: unknown): obj is IMeasureMetadataObject {
    return isMetadataObject(obj) && obj.type === "measure";
}

/**
 * Tests whether the provided object is of type {@link IMeasureMetadataObjectDefinition}.
 *
 * @param obj - object to test
 * @public
 */
export function isMeasureMetadataObjectDefinition(obj: unknown): obj is IMeasureMetadataObjectDefinition {
    return (obj as IMetadataObject).type === "measure" && (obj as IMetadataObject).ref === undefined;
}
