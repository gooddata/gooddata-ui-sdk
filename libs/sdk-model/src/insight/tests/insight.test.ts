// (C) 2019-2020 GoodData Corporation

import {
    IFilter,
    ITotal,
    newAttributeLocator,
    newAttributeSort,
    newMeasureSort,
    newPositiveAttributeFilter,
    newTotal,
    ISortItem,
    IAttributeOrMeasure,
} from "../..";
import { newInsight } from "../../../__mocks__/insights";
import { Account, Activity, Velocity, Won } from "../../../__mocks__/model";
import { IAttribute } from "../../execution/attribute";
import { IBucket, newBucket, BucketItemModifications } from "../../execution/buckets";
import { IMeasure, isMeasure } from "../../execution/measure";
import {
    insightAttributes,
    insightBucket,
    insightBuckets,
    insightHasAttributes,
    insightHasDataDefined,
    insightHasMeasures,
    insightMeasures,
    insightId,
    insightUri,
    insightIsLocked,
    insightCreated,
    insightUpdated,
    insightFilters,
    insightProperties,
    insightSetFilters,
    insightSetProperties,
    insightSetSorts,
    insightSorts,
    insightTotals,
    insightItems,
    VisualizationProperties,
    insightSetBuckets,
    IInsight,
    insightModifyItems,
} from "../index";

const MixedBucket = newBucket("bucket1", Account.Name, Won);
const AttributeBucket = newBucket("bucket2", Activity.Subject);
const MeasureBucket = newBucket("bucket3", Velocity.Sum, Velocity.Min);

const AttributeFilter = newPositiveAttributeFilter(Account.Name, ["myAccount"]);

const VisClassId = "myVis";
const EmptyInsight = newInsight(VisClassId);
const InsightWithSingleBucket = newInsight(VisClassId, (m) => m.buckets([MixedBucket]));
const InsightWithTwoBuckets = newInsight(VisClassId, (m) => m.buckets([AttributeBucket, MixedBucket]));
const InsightWithThreeBuckets = newInsight(VisClassId, (m) =>
    m.buckets([MeasureBucket, AttributeBucket, MixedBucket]),
);
const InsightWithJustAttrBucket = newInsight(VisClassId, (m) => m.buckets([AttributeBucket]));
const InsightWithJustMeasureBucket = newInsight(VisClassId, (m) => m.buckets([MeasureBucket]));

const InvalidScenarios: Array<[string, any]> = [
    ["insight is undefined", undefined],
    ["insight is null", undefined],
];

describe("insightBucket", () => {
    const Scenarios: Array<[string, any, any, IBucket | undefined]> = [
        ["undefined when insight empty", EmptyInsight, undefined, undefined],
        ["undefined when predicate not matched", InsightWithSingleBucket, () => false, undefined],
        ["first found bucket by default", InsightWithTwoBuckets, undefined, AttributeBucket],
        ["bucket by id if predicate is string", InsightWithTwoBuckets, "bucket1", MixedBucket],
    ];

    it.each(Scenarios)("should return %s", (_desc, insightArg, predicateArg, expectedResult) => {
        expect(insightBucket(insightArg, predicateArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightBucket(input)).toThrow();
    });
});

describe("insightBuckets", () => {
    const Scenarios: Array<[string, any, any[], IBucket[]]> = [
        ["empty list if insight empty", EmptyInsight, [], []],
        ["all buckets if no ids provided", InsightWithTwoBuckets, [], [AttributeBucket, MixedBucket]],
        ["bucket matching ID", InsightWithTwoBuckets, ["bucket1"], [MixedBucket]],
    ];

    it.each(Scenarios)("should return %s", (_desc, insightArg, idsArg, expectedResult) => {
        expect(insightBuckets(insightArg, ...idsArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightBuckets(input)).toThrow();
    });
});

describe("insightItems", () => {
    const Scenarios: Array<[string, any, IAttributeOrMeasure[]]> = [
        ["nothing for empty insight", EmptyInsight, []],
        ["items from single bucket insight", InsightWithSingleBucket, [Account.Name, Won]],
        [
            "items from multiple buckets, in correct order",
            InsightWithThreeBuckets,
            [Velocity.Sum, Velocity.Min, Activity.Subject, Account.Name, Won],
        ],
    ];

    it.each(Scenarios)("should return %s", (_desc, insightArg, expectedResult) => {
        expect(insightItems(insightArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightMeasures(input)).toThrow();
    });
});

describe("insightMeasures", () => {
    const Scenarios: Array<[string, any, IMeasure[]]> = [
        ["no measures for empty insight", EmptyInsight, []],
        ["single measure from second bucket", InsightWithTwoBuckets, [Won]],
        [
            "measures in correct order from multiple buckets",
            InsightWithThreeBuckets,
            [Velocity.Sum, Velocity.Min, Won],
        ],
    ];

    it.each(Scenarios)("should return %s", (_desc, insightArg, expectedResult) => {
        expect(insightMeasures(insightArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightMeasures(input)).toThrow();
    });
});

describe("insightHasMeasures", () => {
    const Scenarios: Array<[boolean, string, any]> = [
        [false, "empty insight", EmptyInsight],
        [false, "attr-only insight", InsightWithJustAttrBucket],
        [true, "mixed bucket insight", InsightWithSingleBucket],
        [true, "measure-only insight", InsightWithJustMeasureBucket],
    ];

    it.each(Scenarios)("should return %s for %s", (expectedResult, _desc, insightArg) => {
        expect(insightHasMeasures(insightArg)).toBe(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightHasMeasures(input)).toThrow();
    });
});

describe("insightAttributes", () => {
    const Scenarios: Array<[string, any, IAttribute[]]> = [
        ["no attributes for empty insight", EmptyInsight, []],
        ["single attribute", InsightWithSingleBucket, [Account.Name]],
        [
            "measures in correct order from multiple buckets",
            InsightWithThreeBuckets,
            [Activity.Subject, Account.Name],
        ],
    ];

    it.each(Scenarios)("should return %s", (_desc, insightArg, expectedResult) => {
        expect(insightAttributes(insightArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightAttributes(input)).toThrow();
    });
});

describe("insightHasAttributes", () => {
    const Scenarios: Array<[boolean, string, any]> = [
        [false, "empty insight", EmptyInsight],
        [true, "attr-only insight", InsightWithJustAttrBucket],
        [true, "mixed bucket insight", InsightWithSingleBucket],
        [false, "measure-only insight", InsightWithJustMeasureBucket],
    ];

    it.each(Scenarios)("should return %s for %s", (expectedResult, _desc, insightArg) => {
        expect(insightHasAttributes(insightArg)).toBe(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightHasAttributes(input)).toThrow();
    });
});

describe("insightHasDataDefined", () => {
    const Scenarios: Array<[boolean, string, any]> = [
        [false, "empty insight", EmptyInsight],
        [true, "attr-only insight", InsightWithJustAttrBucket],
        [true, "mixed bucket insight", InsightWithSingleBucket],
        [true, "measure-only insight", InsightWithJustMeasureBucket],
    ];

    it.each(Scenarios)("should return %s for %s", (expectedResult, _desc, insightArg) => {
        expect(insightHasDataDefined(insightArg)).toBe(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightHasDataDefined(input)).toThrow();
    });
});

describe("insightFilters", () => {
    const InsightWithFilters = newInsight(VisClassId, (m) =>
        m.buckets([MixedBucket]).filters([AttributeFilter]),
    );

    const Scenarios: Array<[string, any, IFilter[]]> = [
        ["no filters for empty insight", EmptyInsight, []],
        ["no filters if insight without filters ", InsightWithTwoBuckets, []],
        ["filters from insight", InsightWithFilters, [AttributeFilter]],
    ];

    it.each(Scenarios)("should return %s", (_desc, insightArg, expectedResult) => {
        expect(insightFilters(insightArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightFilters(input)).toThrow();
    });
});

describe("insightSorts", () => {
    const AccountSort = newAttributeSort(Account.Name, "desc");
    const WonSort = newMeasureSort(Won, "asc");
    const WonAndAccountSort = newMeasureSort(Won, "asc", [newAttributeLocator(Account.Name, "myAccount")]);

    const InsightWithValidAttributeSort = newInsight(VisClassId, (m) =>
        m.buckets([MixedBucket]).sorts([AccountSort]),
    );
    const InsightWithValidMeasureSort1 = newInsight(VisClassId, (m) =>
        m.buckets([MixedBucket]).sorts([WonSort]),
    );
    const InsightWithValidMeasureSort2 = newInsight(VisClassId, (m) =>
        m.buckets([MixedBucket]).sorts([WonAndAccountSort]),
    );
    const InsightWithInvalidAttributeSort = newInsight(VisClassId, (m) =>
        m.buckets([MixedBucket]).sorts([newAttributeSort(Activity.Subject)]),
    );
    const InsightWithInvalidMeasureSort1 = newInsight(VisClassId, (m) =>
        m.buckets([MixedBucket]).sorts([newMeasureSort(Velocity.Min)]),
    );
    const InsightWithInvalidMeasureSort2 = newInsight(VisClassId, (m) =>
        m
            .buckets([MixedBucket])
            .sorts([newMeasureSort(Won, "desc", [newAttributeLocator(Activity.Subject, "myActivity")])]),
    );

    const Scenarios: Array<[string, any, ISortItem[]]> = [
        ["no sorts for empty insight", EmptyInsight, []],
        ["no sorts if insight without sorts ", InsightWithTwoBuckets, []],
        ["valid attribute sort", InsightWithValidAttributeSort, [AccountSort]],
        ["valid measure sort without locator", InsightWithValidMeasureSort1, [WonSort]],
        ["valid measure sort with locator", InsightWithValidMeasureSort2, [WonAndAccountSort]],
        ["no sorts for invalid attribute sort", InsightWithInvalidAttributeSort, []],
        ["no sorts for invalid measure sort without locator", InsightWithInvalidMeasureSort1, []],
        ["no sorts for invalid attr locator in measure sort", InsightWithInvalidMeasureSort2, []],
    ];

    it.each(Scenarios)("should return %s", (_desc, insightArg, expectedResult) => {
        expect(insightSorts(insightArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightSorts(input)).toThrow();
    });
});

describe("insightTotals", () => {
    const Total = newTotal("sum", Won, Account.Name);
    const BucketWithTotals = newBucket("bucketWithTotals", Won, Account.Name, Total);
    const InsightWithTotals = newInsight(VisClassId, (m) => m.buckets([BucketWithTotals]));

    const Scenarios: Array<[string, any, ITotal[]]> = [
        ["no totals for empty insight", EmptyInsight, []],
        ["no totals if insight without totals ", InsightWithTwoBuckets, []],
        ["totals from insight", InsightWithTotals, [Total]],
    ];

    it.each(Scenarios)("should return %s", (_desc, insightArg, expectedResult) => {
        expect(insightTotals(insightArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightTotals(input)).toThrow();
    });
});

describe("insightProperties", () => {
    const Properties = { grid: true };
    const InsightWithProperties = newInsight(VisClassId, (m) =>
        m.buckets([MixedBucket]).properties(Properties),
    );

    const Scenarios: Array<[string, any, VisualizationProperties]> = [
        ["no properties for empty insight", EmptyInsight, {}],
        ["no properties if insight without properties ", InsightWithTwoBuckets, {}],
        ["totals from insight", InsightWithProperties, Properties],
    ];

    it.each(Scenarios)("should return %s", (_desc, insightArg, expectedResult) => {
        expect(insightProperties(insightArg)).toEqual(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightProperties(input)).toThrow();
    });
});

describe("insightId", () => {
    it("should return the identifier", () => {
        expect(insightId(EmptyInsight)).toEqual("random");
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightId(input)).toThrow();
    });
});

describe("insightUri", () => {
    it("should return the uri", () => {
        expect(insightUri(EmptyInsight)).toEqual("random");
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightUri(input)).toThrow();
    });
});

describe("insightIsLocked", () => {
    const LockedInsight = newInsight(VisClassId, (m) => m.isLocked(true));

    const Scenarios: Array<[boolean, string, any]> = [
        [false, "non-locked insight", EmptyInsight],
        [true, "locked insight", LockedInsight],
    ];

    it.each(Scenarios)("should return %s for %s", (expectedResult, _desc, insightArg) => {
        expect(insightIsLocked(insightArg)).toBe(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightIsLocked(input)).toThrow();
    });
});

describe("insightCreated", () => {
    const insightCreatedDate = "2020-01-31 13:24:07";
    const SavedInsight = newInsight(VisClassId, (m) => m.created(insightCreatedDate));

    const Scenarios: Array<[string | undefined, string, any]> = [
        [undefined, "insight that has not been saved", EmptyInsight],
        [insightCreatedDate, "saved insight", SavedInsight],
    ];

    it.each(Scenarios)("should return %s for %s", (expectedResult, _desc, insightArg) => {
        expect(insightCreated(insightArg)).toBe(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightCreated(input)).toThrow();
    });
});

describe("insightUpdated", () => {
    const insightUpdatedDate = "2020-01-31 13:24:07";
    const UpdatedInsight = newInsight(VisClassId, (m) => m.updated(insightUpdatedDate));

    const Scenarios: Array<[string | undefined, string, any]> = [
        [undefined, "insight that has not been updated", EmptyInsight],
        [insightUpdatedDate, "updated insight", UpdatedInsight],
    ];

    it.each(Scenarios)("should return %s for %s", (expectedResult, _desc, insightArg) => {
        expect(insightUpdated(insightArg)).toBe(expectedResult);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightUpdated(input)).toThrow();
    });
});

describe("insightSetFilters", () => {
    const AccountFilter = newPositiveAttributeFilter(Account.Name, { values: ["foo"] });
    const ActivityFilter = newPositiveAttributeFilter(Activity.Subject, { values: ["bar"] });

    it("should set new filters in insight without them", () => {
        expect(insightSetFilters(EmptyInsight, [AccountFilter]).insight.filters).toEqual([AccountFilter]);
    });

    it("should overwrite filters in insight that has them", () => {
        const InsightWithSomeFilters = insightSetFilters(EmptyInsight, [AccountFilter]);

        expect(insightSetFilters(InsightWithSomeFilters, [ActivityFilter]).insight.filters).toEqual([
            ActivityFilter,
        ]);
    });

    it("should clear filters if called without parameter", () => {
        const InsightWithSomeFilters = insightSetFilters(EmptyInsight, [AccountFilter]);

        expect(insightSetFilters(InsightWithSomeFilters).insight.filters).toEqual([]);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightSetFilters(input)).toThrow();
    });
});

describe("insightSetSorts", () => {
    const AccountSort = newAttributeSort(Account.Name, "desc");
    const WonSort = newMeasureSort(Won, "asc");

    it("should set new sorts in insight without them", () => {
        expect(insightSetSorts(EmptyInsight, [AccountSort]).insight.sorts).toEqual([AccountSort]);
    });

    it("should overwrite sorts in insight that has them", () => {
        const InsightWithSomeSorts = insightSetSorts(EmptyInsight, [AccountSort]);

        expect(insightSetSorts(InsightWithSomeSorts, [WonSort]).insight.sorts).toEqual([WonSort]);
    });

    it("should clear sorts if called without parameter", () => {
        const InsightWithSomeSorts = insightSetSorts(EmptyInsight, [AccountSort]);

        expect(insightSetSorts(InsightWithSomeSorts).insight.sorts).toEqual([]);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightSetSorts(input)).toThrow();
    });
});

describe("insightSetProperties", () => {
    const SampleProperties1 = { grid: true };
    const SampleProperties2 = { grid: false };

    it("should set new properties in insight without them", () => {
        expect(insightSetProperties(EmptyInsight, SampleProperties1).insight.properties).toEqual(
            SampleProperties1,
        );
    });

    it("should overwrite properties in insight that has them", () => {
        const InsightWithSomeProperties = insightSetProperties(EmptyInsight, SampleProperties1);

        expect(insightSetProperties(InsightWithSomeProperties, SampleProperties2).insight.properties).toEqual(
            SampleProperties2,
        );
    });

    it("should clear properties if called without parameter", () => {
        const InsightWithSomeProperties = insightSetProperties(EmptyInsight, SampleProperties1);

        expect(insightSetProperties(InsightWithSomeProperties).insight.properties).toEqual({});
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightSetProperties(input)).toThrow();
    });
});

describe("insightSetBuckets", () => {
    const SampleBuckets1: IBucket[] = [newBucket("foo", Activity.Default)];
    const SampleBuckets2: IBucket[] = [newBucket("bar", Account.Default)];

    it("should overwrite bucket in an insight", () => {
        const InsightWithSomeBuckets = insightSetBuckets(EmptyInsight, SampleBuckets1);

        expect(insightSetBuckets(InsightWithSomeBuckets, SampleBuckets2).insight.buckets).toEqual(
            SampleBuckets2,
        );
    });

    it("should clear buckets if called without parameter", () => {
        const InsightWithSomeBuckets = insightSetBuckets(EmptyInsight, SampleBuckets1);

        expect(insightSetBuckets(InsightWithSomeBuckets).insight.buckets).toEqual([]);
    });

    it.each(InvalidScenarios)("should throw when %s", (_desc, input) => {
        expect(() => insightSetBuckets(input)).toThrow();
    });
});

describe("insightModifyItems", () => {
    const modifications: BucketItemModifications = (bucketItem: IAttributeOrMeasure): IAttributeOrMeasure => {
        if (isMeasure(bucketItem)) {
            (bucketItem as IMeasure).measure.title = "Modified measure title";
        } else {
            (bucketItem as IAttribute).attribute.alias = "Modified attribute alias";
        }
        return bucketItem;
    };

    it("should return another insight with empty bucket", () => {
        const modifiedInsight: IInsight = insightModifyItems(EmptyInsight, modifications);
        expect(modifiedInsight !== EmptyInsight).toBe(true);
        expect(modifiedInsight.insight.buckets.length).toBe(0);
    });

    it("should return another insight with modified bucket items", () => {
        const insightWithSomeBuckets: IInsight = insightSetBuckets(EmptyInsight, [
            AttributeBucket,
            MeasureBucket,
            MixedBucket,
        ]);
        const modifiedInsight: IInsight = insightModifyItems(insightWithSomeBuckets, modifications);
        expect(modifiedInsight !== insightWithSomeBuckets).toBe(true);
        expect(modifiedInsight.insight.buckets.length).toBe(3);
        expect(modifiedInsight.insight.buckets[0].items.length).toBe(1);
        expect((modifiedInsight.insight.buckets[0].items[0] as IAttribute).attribute.alias).toBe(
            "Modified attribute alias",
        );
        expect(modifiedInsight.insight.buckets[1].items.length).toBe(2);
        expect((modifiedInsight.insight.buckets[1].items[0] as IMeasure).measure.title).toBe(
            "Modified measure title",
        );
        expect((modifiedInsight.insight.buckets[1].items[1] as IMeasure).measure.title).toBe(
            "Modified measure title",
        );
        expect(modifiedInsight.insight.buckets[2].items.length).toBe(2);
        expect((modifiedInsight.insight.buckets[2].items[0] as IAttribute).attribute.alias).toBe(
            "Modified attribute alias",
        );
        expect((modifiedInsight.insight.buckets[2].items[1] as IMeasure).measure.title).toBe(
            "Modified measure title",
        );
    });
});
