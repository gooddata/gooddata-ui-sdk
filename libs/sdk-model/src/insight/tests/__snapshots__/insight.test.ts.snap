// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`insightDisplayFormusage > should return object with empty lists for empty insight 1`] = `
{
  "inAttributes": [],
  "inFilters": [],
  "inMeasureFilters": [],
}
`;
